stan/math/opencl/matrix_cl.hpp

```diff
@@ -294,12 +294,48 @@ class matrix_cl<T, require_arithmetic_t<T>> {
    * @throw <code>std::system_error</code> if the memory on the device could not
    * be allocated
    */
-  template <typename Mat, require_eigen_t<Mat>..., require_same_vt<Mat, T>...>
+  template <typename Mat, require_eigen_t<Mat>..., require_same_vt<Mat, T>...,
+            require_not_t<is_eigen_contiguous_map<Mat>>...>
   explicit matrix_cl(Mat&& A,
                      matrix_cl_view partial_view = matrix_cl_view::Entire)
       : rows_(A.rows()), cols_(A.cols()), view_(partial_view) {
-    initialize_buffer<std::is_rvalue_reference<Mat&&>::value
-                      || !is_eigen_matrix<Mat>::value>(A.eval().data());
+    using Mat_type = std::decay_t<decltype(A.eval())>;
+    if (size() == 0) {
+      return;
+    }
+    if (is_eigen_matrix_or_array<Mat>::value
+        && std::is_lvalue_reference<Mat>::value) {
+      initialize_buffer(A.eval().data());
+    } else {
+      auto* A_heap = new Mat_type(std::move(A.eval()));
+      try {
+        cl::Event e = initialize_buffer(A_heap->data());
+        e.setCallback(CL_COMPLETE, &delete_it<Mat_type>, A_heap);
```

Comment:
I'm not following here, doesn't the initialization already use the heap? Why do we need new and delete here (i.e. explicit mem management over RAII)? We very much try not to do explicit new and deletes in the code so unless there's a big performance gain here (you can use Sean's perf-math to test that) I'd like to just use basic RAII

https://github.com/seantalts/perf-math

Reply:
We also have to be very, very careful about managing resources if we have news floating around. Any possible early exit condition now is a potential leak. 

I was tracing through some of this code; I’d like to make sure it’s tested in exceptional cases so we don’t have leaks.

Reply:
Also though, I think we can use this callback scheme in the error checks so we don't have to stop there!

Reply:
I don't think check* functions should add callbacks just because we have one use case that needs them.

I did not use RAII here because lifetime of the allocated memory is not bound to any object or srtack frame. So using RAII here is not possible. Here I am not allocating memory for the data, just metadata (Matrix object). Actual content of the matrix remains where it is - just the matrix is moved to heap so that it is not destroxed when we return.

I agree we need to be very careful about directley managing memory. I guess you, as reviewer must make sure the code is understandable and correct. In this case I set the callback to delete memory immediately after allocation. If any thing goes wrong and exception is throwed it is deleted in catch body.

I don't think we can test for memory leaks. So understanding the code and making sure it is correct is even more important.

Reply:
I agree that we need to understand how to make this correct, but it’s not the role of the reviewer to figure all of it out. Since this is a change in the paradigm, the submitter really needs to help explain what is happening, why, and what the side effects are. (That’s part of the submission process and documentation of the PR. I’m happy to dig in as the reviewer, but as the submitter, you’ve got to justify this as a maintainable change, which I think it is.)

Just to avoid a situation like other PRs where I’ve tried to help submitters with their PR, I’m inclined to close this PR because it’s actually missing doc around the side effects of this PR. Since this introduces something where the client has to manage memory, it would make sense that the side effect is listed and you’ve outlined under what conditions this has to be cleaned up by the user. 

If you can’t test for leaks, maybe there’s a different design pattern that will allow for testing? Do you want to have a design review? Given the proposed change, we have to comb through all error conditions that could pop up and make sure it’s handled properly. Have you done that? If so, can you highlight how you’ve accomplished that?

(Sorry for so many questions.... these are all things you can help any reviewer with.)

Reply:
This logic is actually pretty confusing. If any of those steps fail, it deletes memory? Why shouldn’t it be on the one call where it transfers the ownership of the memory? If it fails on the second call, why would this function still own the memory?

Reply:
This logic is actually pretty confusing. If any of those steps fail, it deletes memory? Why shouldn’t it be on the one call where it transfers the ownership of the memory? If it fails on the second call, why would this function still own the memory?

Reply:
I'm resolving this conversation. I had to dig further, but I think the logic here isn't quite clear / right. I'll make a new comment.